filters/auth/jwt_validation.go

```diff
@@ -0,0 +1,257 @@
+	infoTemp, ok := ctx.StateBag()[oidcClaimsCacheKey]
+	if !ok {
+		token, ok := getToken(r)
+		if !ok || token == "" {
+			unauthorized(ctx, "", missingToken, f.authClient.url.Hostname(), "")
+			return
+		}
+
+		body, err := f.authClient.getTokeninfo("", ctx)
+		if err != nil {
+			log.Errorf("Error while getting jwt keys: %v.", err)
+
+			unauthorized(ctx, "", "jwt public keys", f.authClient.url.Hostname(), "")
+			return
+		}
+		//var body map[string]interface{}
+		//json.NewDecoder(resp.Body).Decode(&body)
+		rsakeys = make(map[string]*rsa.PublicKey)
+		if body["keys"] != nil {
+			for _, bodykey := range body["keys"].([]interface{}) {
+				key := bodykey.(map[string]interface{})
+				kid := key["kid"].(string)
+				rsakey := new(rsa.PublicKey)
+				number, _ := base64.RawURLEncoding.DecodeString(key["n"].(string))
+				rsakey.N = new(big.Int).SetBytes(number)
+				rsakey.E = 65537
+				rsakeys[kid] = rsakey
+			}
+		} else {
+			log.Error("Not able to get public keys")
+			unauthorized(ctx, "", "Not able to get public keys", f.authClient.url.Hostname(), "")
+			return
+		}
+
+		parsedToken, err := jwt.Parse(token, func(token *jwt.Token) (interface{}, error) {
+			return rsakeys[token.Header["kid"].(string)], nil
+		})
+		if err != nil {
+			log.Errorf("Error while parsing jwt token : %v.", err)
+			unauthorized(ctx, "", "error parsing jwt token", f.authClient.url.Hostname(), "")
+			return
+		} else if !parsedToken.Valid {
+			log.Errorf("Invalid token")
+			unauthorized(ctx, "", "Invalid token", f.authClient.url.Hostname(), "")
+			return
+		} else if parsedToken.Header["alg"] == nil {
+			log.Errorf("alg must be defined")
+			unauthorized(ctx, "", "alg must be defined", f.authClient.url.Hostname(), "")
+			return
+		}
+
+		info = *parsedToken
+		/*err = json.Unmarshal([]byte(parsedToken.Raw), &info)
+		if err != nil {
+			log.Errorf("Error while pasing jwt token : %v.", err)
+			unauthorized(ctx, "", "error parsing jwt token", f.authClient.url.Hostname(), "")
+			return*/
+		//}
+	} else {
+		info = infoTemp.(jwt.Token)
+	}
+
+	/*sub, err := info.Sub()
+	if err != nil {
+		if err != errInvalidTokenintrospectionData {
+			log.Errorf("Error while reading token: %v.", err)
+		}
+
+		unauthorized(ctx, sub, invalidSub, f.authClient.url.Hostname(), "")
+		return
+	}
+
+	if !info.Active() {
+		unauthorized(ctx, sub, inactiveToken, f.authClient.url.Hostname(), "")
+		return
+	}*/
+
+	sub := info.Claims.(jwt.MapClaims)["sub"].(string)
+
+	var allowed = f.validateAnyClaims(info)
+	if !allowed {
+		unauthorized(ctx, sub, invalidClaim, f.authClient.url.Hostname(), "")
+		return
+	}
+
+	authorized(ctx, sub)
+
+	var container2 tokenContainer
+	container2.Claims = make(map[string]interface{})
+	for claim, value := range info.Claims.(jwt.MapClaims) {
+		container2.Claims[claim] = value
+	}
+	//container2.Claims = info.Claims.(jwt.MapClaims)
+
+	log.Errorf("Length %d", len(container2.Claims))
+	ctx.StateBag()[oidcClaimsCacheKey] = container2
```

Comment:
Please do not re-use oidcClaimsCacheKey, but better implement the functionality based on the oidc `oidcClaimsQuery()` filter. As an example that re-uses the code of `oidcClaimsQuery()` see https://github.com/zalando/skipper/blob/master/filters/auth/grantclaimsquery.go for the implementation of the query function.

Reply:
@szuecs I need more clarification on this. grantClaimsQuery re-uses the functionality of oidcClaimsQuery, but I don't see  a place where oidcClaimsCacheKey gets overriden. So if I don't miss anything grantClaimsQuery should get a payload with oidcClaimsCacheKey as well. However I see tokeninfo filter is storing payload using  tokeninfoCacheKey = "tokeninfo"
https://github.com/zalando/skipper/blob/master/filters/auth/tokeninfo.go#L356 and tokenIntrospection is using tokenintrospectionCacheKey   = "tokenintrospection"
https://github.com/zalando/skipper/blob/master/filters/auth/tokenintrospection.go#L473 

How does this work?

Reply:
I think grant filter reuses statebag key
https://github.com/zalando/skipper/blob/39f672857320b501e86149bfc12962c47d4ab761/filters/auth/grant.go#L187-L194
and `grantClaimsQuery` simply creates an instance of `oidcClaimsQuery` under another name https://github.com/zalando/skipper/blob/39f672857320b501e86149bfc12962c47d4ab761/filters/auth/grantclaimsquery.go#L1-L6

We can do the same here if needed or make `oidcIntrospectionSpec` configurable with respect to the statebag key.

Reply:
Guys, still don't get what is the plan. Do I need to introduce new key, adjust oidcIntrospectionSpec to allow key's override and create a new xxxClaimsQuery re-using oidcIntrospectionSpec?

Reply:
IMO we can reuse the key as grant filter does for now.

Reply:
so, I am not changing anything on this